CASSANDRA-19791: Remove other uses of Apache Commons lang for hashcode, equality and r…

jberragan opened the pull request:
…andom string generation

jberragan committed:
Remove other uses of Apache Commons lang for hashcode, equality and random string generation

frankgh reviewed (approved):
+1 Looks good to me

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/RandomUtils.java:
```diff
@@ -63,6 +64,24 @@ public static int randomPositiveInt(int bound)
         return RANDOM.nextInt(bound - 1) + 1;
     }
 
+    public static int nextInt(final int startInclusive, final int endExclusive)
```
please remove `final`

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/RandomUtils.java:
```diff
@@ -63,6 +64,24 @@ public static int randomPositiveInt(int bound)
+            return startInclusive;
+        }
+
+        return startInclusive + RANDOM.nextInt(endExclusive - startInclusive);
```
The bound value for random is exclusive. I think we are missing a value here. 

For example, when `endExclusive - startInclusive == 1`, the produce pseudo random values contains only a single value of 0, but it wants both 0 and 1 instead. 

The util is only used for testing, hence a minor issue. Let's get it correct still.

jberragan replied:
If it is end exclusive and start inclusive and the diff is 1, then shouldn't it only return 1 value? E.g. (4, 5] should always return 4?

yifan-c replied:
Ah. Then the if block above should throw IllegalArgumentException too? e.g. [4, 4) is empty 

```java
        if (startInclusive == endExclusive)
        {
            return startInclusive;
        }
```

jberragan replied:
It currently does this:

```
if (startInclusive == endExclusive)
{
    return startInclusive;
}
```

I don't particularly mind either way, current approach seems more permissive.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/RandomUtils.java:
```diff
@@ -100,8 +119,18 @@ public static InetAddress randomInet()
+    public static String randomAlphanumeric(int maxLength)
     {
-        return type.randomValue(MIN_COLLECTION_SIZE);
+        StringBuilder sb = new StringBuilder(maxLength);
+        IntStream.rangeClosed(0, maxLength)
+                 .map(i -> RANDOM.nextInt(ALPHANUMERIC.length()))
+                 .mapToObj(ALPHANUMERIC::charAt)
+                 .forEach(sb::append);
+        return sb.toString();
```
The alternative in my mind is random from [0, 36), and map the lower 10 to numbers and upper 26 to chars. Therefore, you do not need to look up the ALPHANUMERIC string. 
It is just a nit, feel free to ignore.

jberragan committed:
Review comments
Add randomAsciiAlphanumeric to randomly generate ascii number of uppercase character
Add RandomUtilsTest to verify behavior of RandomUtils

jberragan committed:
Add Apache header and fix checkstyle error

jberragan committed:
Make nextInt require start < end

jberragan committed:
Update CHANGES.txt
